docs/adr/0004-instrumentation.md

```diff
@@ -0,0 +1,99 @@
+# Implement instrumentation via a new API Endpoint with external ad hoc probe calculation
+
+- Status: proposed
+- Deciders: Christina Harlow, Linh Nguyen, Rob Miller
+- Date: [2022-04-05
+
+Technical Story: https://github.com/mozilla/glam/issues/1847
+
+## Context and Problem Statement
+
+In order to help us with various product decisions and overall product quality,
+we want to understand how GLAM is normally used: Where does it fit in a typical
+user's workflow (a starting point, an endpoint, or something in between)?
+
+We can accomplish this by gathering data such as how a typical user session
+looks like, most and least looked at probes, how probe browsing behaves over
+short and long periods of time and whether it correlates with other events such
+as certain product releases.
+
+Additionally, as a more time-sensitive issue, GLAM ETL does a lot of work and
+takes over 12 hours to complete. We intend to reduce GLAM ETL time by finding
+the probes that are most likely to be requested by users from the aforementioned
+data and have the ETL build only those probes. This document also presents a
+plan for ad hoc probe building, which needs to happen when someone requests a
+probe that had not been pre-built by the ETL.
+
+## Decision Drivers
+
+- Usage metrics collection must be reliable and cannot be bypassed or influenced
+  by other systems such as browsers' adblocks
+- GLAM ETL should consume this service so it can optimize its work, only
+  processing probes that are likely to be looked at
+- GLAM App should consume this service so it can trigger an ad hoc calculation
+  whenever a probe that was not processed by GLAM ETL is requested
+- Ad hoc probe building should be independent of GLAM App because it's likely to
+  be CPU/memory intensive, with different needs in terms of scalability
+
+## Considered Options
+
+- 1: Collect metrics on the front end (via a library such as Open Telemetry
+  Instrumentation or a custom-built solution) and send them to the GLAM back end
+  server to be stored in a table in the GLAM App Postgres DB. Expose meaningful
+  api endpoints for usaga data consumption by GLAM App and and GLAM ETL. This
+  provides flexibility to adapt to future requirements and "immunity" to
+  adblocks, but might be longer to implement
+- 2: Use Gleanjs, a JS library for telemetry, which would be quicker than a
+  custom implementation. It's also good because of dogfooding and internal
+  support. The downside here is it can be bypassed by adblockers, because it
+  sends data to a different endpoint
+- 3: Google analytics: currently available, it's quick and easy to implement,
+  but doesn't give us the insight into e.g. session-scoped data and can also be
```

Comment:
FWIW, for this service, the adblocker issue could be solved via cloud functions as reported by Arkadiusz

Reply:
Interesting. IDK much about GA, and didn't know we could specify a custom endpoint to get the metrics sent to.
I did a quick look around and found [this article](https://www.simoahava.com/gtm-tips/send-google-analytics-requests-custom-endpoint/) explaining how to do so.

My thoughts now:
* In case we move fwd with GA, we'll probably want to implement support to [Measurement Protocol](https://developers.google.com/analytics/devguides/collection/protocol/v1) in the backend in order to process those metrics. This can be a good thing because it can serve as a guide, but can also be a burden if it has a lot more than what we need.
* Assuming we don't want to change the way GA currently works for GLAM, what would our options be?

Reply:
Would Glean fit the bill here?

Reply:
Considering the clarifications you provided [here](https://github.com/mozilla/glam/pull/1910#discussion_r852700999) (thanks a lot!) I don't see why it wouldn't